packages/gatsby/src/utils/create-node-id.ts

```diff
@@ -3,8 +3,29 @@ import report from "gatsby-cli/lib/reporter"
 
 const seedConstant = `638f7a53-c567-4eca-8fc1-b23efb1cfb2b`
 
+// This cache prevents duplicate calls to uuid which is relevant for certain cases
+const unprefixedCache: Map<string, string> = new Map()
+const namespacedCache: Map<string, Map<string, string>> = new Map([
+  [``, unprefixedCache],
+]) // ns -> name -> id
+
 /**
- * createNodeId() Generate UUID
+ * Generate a unique id that is consistent, deterministic, and fast while resulting in predictably short hashes.
+ *
+ * Some characteristics for this id:
+ *
+ * - The value of the `id` should not mean anything (it is "ours")
+ * - The value does not need to be encrypted
+ * - The value must be unique within our system (as little collision risk as possible on small ascii inputs)
+ * - The value needs to be deterministic (same input always results in same output)
+ * - The conversion needs to be fast (we used to use uuid, which called into crytpo for sha1, but it was super slow)
```

Comment:
This also could get some re-wording as we do continue to use `uuid`

Reply:
I basically only removed the mention of "used to" because the rest is still true, even if we do use a crypto level hash for now.